### ISSUE TITLE
nexusformat does not open ESRF files

### ISSUE THREAD

**Issue by jkotan** [nexpy/nexusformat#213]
Running nxinspect I found it crashes on on ESRF files i.e.

```
jkotan@dellek:~/sources/nxvalidate$ nxinspect -d ~/sources/definitions -f ~/files/esrf_test.h5 
NXroot: /
    "@create_version" is not defined as an attribute in NXroot
    "@publisher_version" is not defined as an attribute in NXroot
    NXentry: /3.1
        Field: /3.1/end_reason
            This field is not defined in NXentry
Traceback (most recent call last):
  File "/home/jkotan/testinst/local/lib/python3.11/dist-packages/nexusformat/nexus/tree.py", line 4550, in __getitem__
    node = node.entries[name]
           ~~~~~~~~~~~~^^^^^^
KeyError: 'signal'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jkotan/testinst/local/lib/python3.11/dist-packages/nexusformat/nexus/tree.py", line 5291, in __getattr__
    return getattr(self.nxlink, name)
                   ^^^^^^^^^^^
  File "/home/jkotan/testinst/local/lib/python3.11/dist-packages/nexusformat/nexus/tree.py", line 5393, in nxlink
    self._link = self.internal_link
                 ^^^^^^^^^^^^^^^^^^
  File "/home/jkotan/testinst/local/lib/python3.11/dist-packages/nexusformat/nexus/tree.py", line 5425, in internal_link
    return self.nxroot[self._target]
           ~~~~~~~~~~~^^^^^^^^^^^^^^
  File "/home/jkotan/testinst/local/lib/python3.11/dist-packages/nexusformat/nexus/tree.py", line 4552, in __getitem__
    raise NeXusError("Invalid path")
nexusformat.nexus.tree.NeXusError: Invalid path

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jkotan/testinst/local/bin/nxinspect", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/jkotan/testinst/local/lib/python3.11/dist-packages/nxvalidate/scripts/nxinspect.py", line 75, in main
    validate_file(filename, path=path, definitions=definitions)
  File "/home/jkotan/testinst/local/lib/python3.11/dist-packages/nxvalidate/validate.py", line 713, in validate_file
    validator.validate(path)
  File "/home/jkotan/testinst/local/lib/python3.11/dist-packages/nxvalidate/validate.py", line 691, in validate
    validator.validate(item, indent=indent)
  File "/home/jkotan/testinst/local/lib/python3.11/dist-packages/nxvalidate/validate.py", line 363, in validate
    field_validator.validate(tag, item, parent=self, indent=indent)
  File "/home/jkotan/testinst/local/lib/python3.11/dist-packages/nxvalidate/validate.py", line 605, in validate
    self.check_type(field, tag['@type'])
  File "/home/jkotan/testinst/local/lib/python3.11/dist-packages/nxvalidate/validate.py", line 428, in check_type
    if is_valid_number(field.dtype):
                       ^^^^^^^^^^^
  File "/home/jkotan/testinst/local/lib/python3.11/dist-packages/nexusformat/nexus/tree.py", line 5293, in __getattr__
    raise AttributeError(
AttributeError: Cannot resolve the link to '/signal'

```
The file can be found at https://github.com/jkotan/file-examples/blob/master/esrf/esrf_test.h5

**Comment by rayosborn** [comment]
Sorry to take a while to look at this, but thank you for raising the issue. This is a bug that needs fixing. When soft links are created by the nexusformat API, their target paths are automatically converted to absolute paths. Your files contain target paths that are relative. This is valid in HDF5 and is reasonably easy to fix in nexusformat. The only problem is that to maintain backward compatibility with scripts that assumed the paths were absolute, including one of the pytests, I can't just assume that paths are either absolute or relative. 

For example, in the past, someone could use the following in a script:
```
with nxopen('/Users/rosborn/Desktop/test.nxs', 'w') as root:
    root['entry'] = NXentry()
    root['entry/data'] = np.linspace(0,10,11)
    root['entry/link'] = NXlink('entry/data')
```
This would produce the following tree:
```
>>> print(root.tree)
root:NXroot
  entry:NXentry
    data = float64(11)
    link -> /entry/data
```
To fix this issue, it is clear that I should leave the link path as relative, but I will have to catch cases where  the path is relative to the root. I think the following code should be backwardly compatible.
```
class NXlink(NXobject):
    def internal_link(self):
        if Path(self._target).is_absolute():
            return self.nxroot[self._target]
        else:
            try:
                return self.nxgroup[self._target]
            except NeXusError:
                return self.nxroot[self._target]
```
@jkotan, I will produce a PR that contains this fix. If you are able to test it, I will let you know when the PR is ready.

**Comment by rayosborn** [comment]
On further reflection, there are some issues that may need clarifying. The soft links in esrf_test.h5 are to objects in the same directory. This seems like a slightly odd use case, but one that we can support. I suppose there could be instances where the link is to an object in a child group. However, I'm fairly sure that NeXus (and possibly even HDF5) does not support relative paths like '../group/field'. @jkotan, could ESRF files contain those kinds of relative paths?

**Comment by rayosborn** [comment]
I believe this is fixed by PR #214.
